 # Titanic Dataset

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Data Loading:
    The Titanic dataset is available in the Seaborn library, so we can load it directly into a DataFrame:

In [2]:
import pandas as pd

df = pd.read_csv("Titanic data.csv")


In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,0,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,1,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,0,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,0,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,1,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


In [4]:
df.shape #Row and Column

(418, 12)

### The dataset contains 418 entries and 12 columns.Each row represents an individual passenger.
    Survived: The target variable (0= No, 1= Yes).
     Pclass: Passenger Class (1 = Upper, 2= Middle, 3= Lower).
    Sex: Passenger's gender.
    Age: Age in years.
    SibSp / Parch: Indicates family members aboard (Siblings/Spouses/Parents/Children).
    Fare: The price paid for the ticket.
    Embarked: The port of boarding (C, Q, or S).
    C = Cherbourg, Q = Queenstown, S = Southampton)

# Insights:

    - Female passengers had a higher survival rate than male passengers.
    - First Class passengers were more likely to survive compared to passengers in other classes.
    - Age was an important factor, and children had a better chance of survival.
    -Passengers who paid higher fares had a greater probability of survival.
    -Handling missing values was very important for accurate data analysis.

### 3. Exploring the Data:

In [5]:
#To know the information about each column like the data type, etc we use the df.info() function.
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    object 
 4   Sex          418 non-null    object 
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    object 
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     object 
 11  Embarked     418 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


In [6]:
df["Sex"]

0        male
1      female
2        male
3        male
4      female
        ...  
413      male
414    female
415      male
416      male
417      male
Name: Sex, Length: 418, dtype: object

In [7]:
df["Age"].fillna(df["Age"].mean(), inplace=True) #fillna=Missing value 

C:\Users\HP\AppData\Local\Temp\ipykernel_8988\1318142762.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Age"].fillna(df["Age"].mean(), inplace=True) #fillna=Missing value


In [8]:
##Now let's see if there are any NULL values present in the dataset. This can be checked using the isnull() function. It yields the following output.
df.isnull().sum() #Missing Values Check

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [9]:
df.isna()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,True,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,True,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
413,False,False,False,False,False,False,False,False,False,False,True,False
414,False,False,False,False,False,False,False,False,False,False,False,False
415,False,False,False,False,False,False,False,False,False,False,True,False
416,False,False,False,False,False,False,False,False,False,False,True,False


In [10]:
df = df.drop(columns=['Cabin'])

In [11]:
df = df.drop(columns=['Cabin'])
df = df.drop(columns=['Ticket'])
df = df.drop(columns=['PassengerId'])
df = df.drop(columns=['Name'])

KeyError: "['Cabin'] not found in axis"

In [ ]:
df["Age"].fillna(df["Age"].mean(), inplace=True) #fillna=Missing value 

In [ ]:
# Get summary statistics of numerical variables
df.describe()

 ## 4. Data Cleaning:

#### Data cleaning is an essential step in EDA. We must handle missing values, outliers, and inconsistencies in the dataset. Some common data-cleaning tasks include:

In [ ]:
df.duplicated().sum()

In [ ]:
df.Age.mean()
round(df.Age.mean(),1)

In [ ]:
df['Age'] = df['Age'].fillna(round(df['Age'].mean(), 1))


In [ ]:
df[df.Fare.isna()]

In [ ]:
df.dropna(inplace=True)# Remove rows with missing values

In [ ]:
df.head()

In [ ]:
df[df['Fare'].isna()]

In [ ]:
df.head(2)

In [ ]:
df.info()

In [ ]:
df.head()

In [ ]:
# Exploratory Data Analysis (EDA)
df["Survived"] #I analyzed the survival distribution

In [ ]:
df["Survived"].value_counts()

In [ ]:
df.groupby("Sex")["Survived"].mean() # 2. Survival Rate by Gender

In [ ]:
df.

In [ ]:
print("Step: Calculating Survival Rate by Gender...\n")
gender_survival = df.groupby("Sex")["Survived"].mean()
print("Survival Rate (Male vs Female):")
print(gender_survival)
print("\nStep: Plotting Bar Chart...\n")
gender_survival.plot(kind="bar")
plt.title("Survival Rate by Gender")
plt.xlabel("Gender")
plt.ylabel("Survival Rate")
plt.show()

 ### Females had a higher survival probability

# Survival Rate by Passenger Class

In [ ]:
df.groupby("Pclass")["Survived"].mean()

In [ ]:
print("Step: Calculating Survival Rate by Passenger Class...\n")
class_survival = df.groupby("Pclass")["Survived"].mean()
print("Survival Rate by Class (1st, 2nd, 3rd):")
print(class_survival)
print("\nStep: Plotting Bar Chart...\n")
class_survival.plot(kind="bar")
plt.title("Survival Rate by Passenger Class")
plt.xlabel("Passenger Class")
plt.ylabel("Survival Rate")
plt.show()

 ### First class passengers survived more

# Age Distribution

In [ ]:
print("Step: Analyzing Age Distribution...\n")
print("Age Column Summary:")
print(df["Age"].describe())
print("\nStep: Plotting Age Distribution Histogram...\n")
df["Age"].hist()
plt.title("Age Distribution of Passengers")
plt.xlabel("Age")
plt.ylabel("Number of Passengers")
plt.show()

In [ ]:
print("Step: Calculating Correlation Matrix...\n")
corr_matrix = df.corr(numeric_only=True)
print("Correlation Matrix:")
print(corr_matrix)
print("\nStep: Plotting Correlation Heatmap...\n")
sns.heatmap(corr_matrix, annot=True)
plt.title("Correlation Heatmap")
plt.show()

 # 5. Data Visualization:(Plotting Distribution Plots)


### Visualization helps us understand the data and identify patterns. We can create various types of plots using libraries like Matplotlib and Seaborn.

In [ ]:
###CountPlot
sns.catplot(data=df, x="Sex",hue="Survived",kind="count")

In [ ]:
sns.countplot(data=df, x='Pclass', hue='Survived')
plt.title('Survival by Passenger Class')
plt.savefig('class_survival.png')

In [ ]:
# Divide Fare into 4 bins
df['Fare'] = pd.qcut(df['Fare'], 4)

# Barplot - Shows approximate values based 
# on the height of bars.
sns.barplot(x ='Fare', y ='Survived', data = df)

In [ ]:
# Countplot
sns.catplot(x ='Embarked', hue ='Survived', 
kind ='count', col ='Pclass', data = df)

In [ ]:
# Group the dataset by Pclass and Survived and then unstack them
group = df.groupby(['Pclass', 'Survived'])
pclass_survived = group.size().unstack()

# Heatmap - Color encoded 2D representation of data.
sns.heatmap(pclass_survived, annot = True, fmt ="d")

In [ ]:
sns.barplot(data=df,x="Sex",y="Age")

In [ ]:
plt.figure(figsize=(10, 8))
sns.heatmap(df.corr(numeric_only=True), annot=True, cmap="YlGnBu")
plt.title('Correlation Heatmap')
plt.show()


In [ ]:
df.head()

In [ ]:
df

In [ ]:
sns.set_style('whitegrid')
sns.countplot(data=df,x="Survived")

In [ ]:
sns.set_style('whitegrid')
sns.countplot(data=df,x="Survived",hue='Sex',palette='RdBu_r')

In [ ]:
sns.set_style('darkgrid')
sns.countplot(data=df,x='Survived',hue='Pclass',palette='viridis')